src/order/upper_lower.lean

```diff
@@ -515,3 +537,119 @@ def Iic_Inf_hom : Inf_hom α (lower_set α) := ⟨Iic, λ s, (Iic_Inf s).trans I
 
 end complete_lattice
 end lower_set
+
+section closure
+variables [preorder α] {s t : set α} {x : α}
+
+/-- The greatest upper set containing a given set. -/
+def upper_closure (s : set α) : upper_set α :=
+⟨{x | ∃ a ∈ s, a ≤ x}, λ x y h, Exists₂.imp $ λ a _, h.trans'⟩
+
+/-- The least lower set containing a given set. -/
+def lower_closure (s : set α) : lower_set α :=
+⟨{x | ∃ a ∈ s, x ≤ a}, λ x y h, Exists₂.imp $ λ a _, h.trans⟩
+
+@[simp, norm_cast] lemma coe_upper_closure (s : set α) :
+  ↑(upper_closure s) = {x | ∃ a ∈ s, a ≤ x} := rfl
+
+@[simp, norm_cast] lemma coe_lower_closure (s : set α) :
+  ↑(lower_closure s) = {x | ∃ a ∈ s, x ≤ a} := rfl
+
+@[simp] lemma mem_upper_closure : x ∈ upper_closure s ↔ ∃ a ∈ s, a ≤ x := iff.rfl
+@[simp] lemma mem_lower_closure : x ∈ lower_closure s ↔ ∃ a ∈ s, x ≤ a := iff.rfl
+
+lemma subset_upper_closure : s ⊆ upper_closure s := λ x hx, ⟨x, hx, le_rfl⟩
+lemma subset_lower_closure : s ⊆ lower_closure s := λ x hx, ⟨x, hx, le_rfl⟩
```

Comment:
Can we have the minimality property of `lower_closure` in a definitionally nice form? Or does that already come from `gi_lower_closure_coe`